### DISCUSSION TITLE
ReadProcessMemory with nuint

### DISCUSSION THREAD

**Discussion by Lightczx** [microsoft/CsWin32#945] [Q&A] [answered]
The CsWin32 Generated ReadProcessMemory signature is:
```c#
[DllImport("KERNEL32.dll", ExactSpelling = true, SetLastError = true)]
[DefaultDllImportSearchPaths(DllImportSearchPath.System32)]
[SupportedOSPlatform("windows5.1.2600")]
internal static extern unsafe winmdroot.Foundation.BOOL ReadProcessMemory(winmdroot.Foundation.HANDLE hProcess, void* lpBaseAddress, void* lpBuffer, nuint nSize, [Optional] nuint* lpNumberOfBytesRead);
```

While I created wrapper method to read out a uintptr from process memory:

```c#
private static unsafe bool UnsafeReadProcessMemory(Process process, nuint baseAddress, out nuint value)
{
    byte[] buffer = new byte[8];
    BOOL result;
    fixed (byte* pBuffer = buffer)
    {
        result = ReadProcessMemory((HANDLE)process.Handle, (void*)baseAddress, &pBuffer, 8, default);
    }

    value = (nuint)BinaryPrimitives.ReadUInt64LittleEndian(buffer);
    return result;
}
```

It always read 0 and error code is also 0
But when I use:

```c#
[DllImport("kernel32.dll", EntryPoint = "ReadProcessMemory", SetLastError = true)]
private static extern bool SimpleReadProcessMemory(IntPtr hProcess, UIntPtr lpBaseAddress, [Out] byte[] lpBuffer, int dwSize, out IntPtr lpNumberOfBytesRead);
```
and a wrapper method:
```
private static unsafe bool SimpleReadProcessMemoryWrapper(Process process, nuint baseAddress, out nuint value)
{
    byte[] buffer = new byte[8];
    bool result = SimpleReadProcessMemory(process.Handle, baseAddress, buffer, 8, out _);
    value = (nuint)BinaryPrimitives.ReadUInt64LittleEndian(buffer);
    return result;
}
```

It give me a success result.
My program is x64 and the size of uintptr is 8.
Did I miss something?

**Comment by AArnott** [comment, marked as answer]
I notice the `dwSize` parameter type in your working example is declared as `int` in your 64-bit process, but the docs and headers and win32metadata suggest it should be pointer-sized (i.e. `nuint`). That's the only significant difference between your declaration and the one from CsWin32. And I'd lean toward CsWin32's being the correct one, so I'm surprised by your report that yours works and Cswin32's doesn't.

I don't know enough about how to use this function to test it locally however. Can you provide a full console app that perhaps spawns notepad.exe and reads anything from it so I can experiment?

**Reply by Lightczx** [reply]
My bad. It seems like it's uint size issue,after i pass the correct uint size, it works.
